### PR TITLE
Add server pruning

### PR DESCRIPTION
**Pull request by binaryoverload**
- [x] Completed PR - Has all features or fixes that it was meant to and isn't half done
- [x] Follows the style guide - Style guide found [here](https://github.com/FlareBot/FlareBot/blob/master/CONTRIBUTING.md)
- [x] Tested changes

## What is added
Prune Command
Util to handle confirming actions

### CODE REVIEW

File: src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public String getDescription() {
+        return null;
+    }
+
+    @Override
+    public String getUsage() {
+        return "";
+    }
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: Incomplete

File: src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    public String getCommand() {
+        return "prune";
+    }
+
+    @Override
+    public String getDescription() {
+        return null;
+    }
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: Incomplete

File: src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+                    ConfirmLib.pushAction(sender.getId(),
+                            new RestActionWrapper(guild.getGuild().getController()
+                                    .prune(amount)
+                                    .reason("Pruned by user: " + MessageUtils.getTag(sender)), this.getClass()));
+                    return;
+
+                } else {
+                    MessageUtils.sendErrorMessage("You need the permission `flarebot.prune.server` to do this!", channel);
+                }
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: You do realize that the command system does this automatically with getPermission right?

File: src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+public class RestActionWrapper {
+
+    private final RestAction action;
+    private final Class<? extends Command> origin;
+
+    public RestActionWrapper(RestAction action, Class<? extends Command> origin) {
+
+        this.action = action;
+        this.origin = origin;
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java]: Why the whitespace? Why a class and not an instance?

File: src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private static final ExpiringMap<String, List<RestActionWrapper>> confirmMap = new ExpiringMap<>(TimeUnit.MINUTES.toMillis(1));
+
+    public static void pushAction(String userID, RestActionWrapper action) {
+        if (confirmMap.containsKey(userID)) {
+            List<RestActionWrapper> actions = confirmMap.get(userID);
+            if (actions.stream().filter(wrapper -> wrapper.getOrigin().equals(action.getOrigin())).count() == 0) {
+                actions.add(action);
+            }
+        } else {
+            ArrayList<RestActionWrapper> actions = new ArrayList<>();
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java]: Use a concurrent hash set (`ConcurrentHashMap.newKeySet()`) for thread safety and speed and duplication checking

File: src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public RestAction getAction() {
+        return action;
+    }
+
+    public Class<? extends Command> getOrigin() {
+        return origin;
+    }
+}
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java]: Implement hashCode() and equals() (Press Alt+Insert in intellij)

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 7]
```diff
@@ -117,4 +121,68 @@ public static String truncate(int length, String string, boolean ellipse) {
...
+    public static List<Role> getRole(String string, Guild guild) {
+        return guild.getRolesByName(string, true);
+    }
+
+    public static User getUser(String s) {
+         return getUser(s, "");
+    }
+
+    public static User getUser(String s, String guildId) {
+        if (GeneralUtils.userDiscrim.matcher(s).find()) {
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Why do you do `GeneralUtils.`?

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 8]
```diff
@@ -6,24 +6,27 @@
...
 import java.time.format.DateTimeFormatter;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.ExecutionException;
+import java.util.regex.Pattern;
 
 public class GeneralUtils {
 
     private static final DecimalFormat percentageFormat = new DecimalFormat("#.##");
+    private static final Pattern userDiscrim = Pattern.compile(".#[0-9]{4}");
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Will not work

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 9]
```diff
@@ -117,4 +121,68 @@ public static String truncate(int length, String string, boolean ellipse) {
...
+            } catch (NumberFormatException | NullPointerException ignored) {
+            }
+        }
+        return null;
+    }
+
+    public static Role getRole(String s, String guildId) {
+        Role role;
+        role = FlareBot.getInstance().getGuildByID(guildId).getRoles().stream()
+                .filter(rolePredicate -> rolePredicate.getName().equalsIgnoreCase(s))
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Why call it rolePredicate?

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 10]
```diff
@@ -117,4 +121,68 @@ public static String truncate(int length, String string, boolean ellipse) {
...
+                }
+                if (tmp != null) return tmp;
+            } catch (NumberFormatException | NullPointerException ignored) {
+            }
+        }
+        return null;
+    }
+
+    public static Role getRole(String s, String guildId) {
+        Role role;
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Why not inline this?

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 11]
```diff
@@ -117,4 +121,68 @@ public static String truncate(int length, String string, boolean ellipse) {
...
+        role = FlareBot.getInstance().getGuildByID(guildId).getRoles().stream()
+                .filter(rolePredicate -> rolePredicate.getName().equalsIgnoreCase(s))
+                .findFirst().orElse(null);
+        if (role != null) return role;
+        try {
+
+            role = FlareBot.getInstance().getGuildByID(guildId).getRoleById(Long.parseLong(s.replaceAll("[^0-9]", "")));
+
+            if (role != null) return role;
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: What's it with you and blank lines?

File: src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+            actions.add(action);
+            confirmMap.put(userID, actions);
+        }
+    }
+
+    public static RestAction get(String userID, Class<? extends Command> command) {
+        if (confirmMap.containsKey(userID)) {
+            List<RestActionWrapper> wrappers = confirmMap.get(userID);
+            return ((RestActionWrapper) wrappers.stream().filter(wrapper -> wrapper.getOrigin().equals(command)).toArray()[0]).getAction();
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java]: `findFIrst()`

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java [hunk 13]
```diff
@@ -117,4 +121,68 @@ public static String truncate(int length, String string, boolean ellipse) {
         return string.substring(0, Math.min(string.length(), length)) + (string.length() > length ? "..." : "");
     }
 
+    public static List<Role> getRole(String string, Guild guild) {
+        return guild.getRolesByName(string, true);
+    }
+
+    public static User getUser(String s) {
+         return getUser(s, "");
```
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Use `null` instead

File: src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java (hunk 6 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/objects/RestActionWrapper.java]: kk

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 9 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: role was taken, autocompleted by intellij and it made sense

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 13 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Not my code originally but OK

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 9 above)
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: I'd do `r` to be honest.

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 13 above)
Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: I wonder whose it was, but do it anyways

File: src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java (hunk 5 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/ConfirmLib.java]: K

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 8 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Ah yes, I see the issue. Will fix

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 7 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: WIll fix

File: src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java (hunk 10 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/util/GeneralUtils.java]: Fixed

File: src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java (hunk 3 above)
Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: I don't believe it does. I believe it only does this for the root command itself

Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: @ArsenArsen Double checked, don't think it does

Comment (ArsenArsen) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: Of course, but why would you want anyone to be confirming if they can't get to the confirmation anyway?

Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: They can?

Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: ooo

Comment (binaryoverload) [inline src/main/java/stream/flarebot/flarebot/commands/moderation/PruneCommand.java]: didn't return, rip

**Comment by binaryoverload** [comment]
@ArsenArsen We done here?

**Comment by ArsenArsen** [comment]
No, it still does not use the build in permission API

**Pull request merged**